### PR TITLE
Refactor EventHubDataAdapter to be plugable

### PR DESCRIPTION
**Pull request by jason-bragg**
Attempting to address "Stream data adapter #4621" for EventHub streams.
EventHubDataAdapter is now pluggable.  It can now be configured without customizing queue adapter or cache.

Other changes:
- Simplified pooled cache to only use CachedMessage
- Simplified eviction strategy to only use CachedMessage
- Removed the need for custom comparer logic.
- Added interned stream namespace to CachedMessage to prevent the need to encode it into pooled memory segment.
- Refactored IQueueDataAdapter to support common send behaviors.

Overall, pooled cache pattern is much cleaner and maintainable, at the cost of some customizability.

### CODE REVIEW

File: src/Azure/Orleans.Streaming.AzureStorage/Providers/Streams/AzureQueue/IAzureQueueDataAdapter.cs [hunk 1]
```diff
@@ -73,7 +73,7 @@ public AzureQueueDataAdapterV2(SerializationManager serializationManager)
         /// <summary>
         /// Creates a cloud queue message from stream event data.
         /// </summary>
-        public CloudQueueMessage ToQueueMessage<T>(Guid streamGuid, string streamNamespace, IEnumerable<T> events, Dictionary<string, object> requestContext)
+        public CloudQueueMessage ToQueueMessage<T>(Guid streamGuid, string streamNamespace, IEnumerable<T> events, StreamSequenceToken token, Dictionary<string, object> requestContext)
```
Comment (xiazen) [inline src/Azure/Orleans.Streaming.AzureStorage/Providers/Streams/AzureQueue/IAzureQueueDataAdapter.cs]: Is this AzureQueueDataAdaptor related change to this pr?

File: src/Orleans.Core.Abstractions/Streams/Core/StreamSequenceToken.cs [hunk 2]
```diff
@@ -10,6 +10,16 @@ namespace Orleans.Streams
     [Serializable]
     public abstract class StreamSequenceToken : IEquatable<StreamSequenceToken>, IComparable<StreamSequenceToken>
     {
+        /// <summary>
+        /// Number of event batches in stream prior to this event batch
+        /// </summary>
+        public long SequenceNumber { get; protected set; }
```
Comment (xiazen) [inline src/Orleans.Core.Abstractions/Streams/Core/StreamSequenceToken.cs]: what does this mean to other persistent stream providers?

Comment (xiazen) [inline src/Orleans.Core.Abstractions/Streams/Core/StreamSequenceToken.cs]: what does this mean to other stream providers which doesn't support rewindability

File: src/Azure/Orleans.Streaming.AzureStorage/Providers/Streams/AzureQueue/IAzureQueueDataAdapter.cs (hunk 1 above)
Comment (jason-bragg) [reply src/Azure/Orleans.Streaming.AzureStorage/Providers/Streams/AzureQueue/IAzureQueueDataAdapter.cs]: Yes, unfortunately.  The interface surface changed, thus so need the implementations.

File: src/Orleans.Core.Abstractions/Streams/Core/StreamSequenceToken.cs (hunk 2 above)
Comment (jason-bragg) [reply src/Orleans.Core.Abstractions/Streams/Core/StreamSequenceToken.cs]: It means that ordering information needs be expressed in a long value.

This could be a problem if systems have more complex or convoluted ordering information.

It could also be an issue if provider implementations have persisted tokens that don't fit this.  They may need to perform some translation between the old token and a new compatible one.  This should not be an issue for any we maintain, but custom tokens may encounter some issues.

Comment (jason-bragg) [reply src/Orleans.Core.Abstractions/Streams/Core/StreamSequenceToken.cs]: It shouldn't matter for those, as they shouldn't rely on tokens.  In those cases, this information is only used to order data in the cache for delivery.

Comment (jason-bragg) [reply src/Orleans.Core.Abstractions/Streams/Core/StreamSequenceToken.cs]: The cache ordering requirement already existed even for non-rewindable stream providers, thus this is not a new constraint, it was just not expressed in the token.

**Comment by xiazen** [comment]
@jason-bragg LGTM. do you want to run it by back-compact test, and if it went ok, we merge it?
